Multi-tiered masternode update

kemi04 opened the pull request:
Win counting functionality moved to ProcessNewBlock in main.cpp file. This ensures that all masternodes keep track of masternode wins.

kemi04 committed:
Multi-tier masternodes fixed

kemi04 committed:
Multi-tier functionality added

Multi-tier functionality added

kemi04 committed:
Multi-tier masternode functionality fixed

kemi04 committed:
Indentation fixed

kemi04 committed:
Multi-tier functionality update

kemi04 committed:
Multi-tier masternode patch

This patch should prevent doulbe tier-1 masternode payments for two consecutive blocks.
Multi-tier masternode last-paid timer has been updated to use blocks for last time paid calculation.
Other minor changes.

kemi04 committed:
Multi-tier masternode patch-2

This update should prevent multi-tier masternodes from winning more times than needed per cycle and fix the GetLastPaid() method.

kemi04 committed:
Reversed accidental changes

kemi04 committed:
Multi-tier masternode patch-3

Instead of looking at the numbers of wins of masternode objects, the number of wins will be calculated by checking the block chain winners, as double payments occur sometimes.
Last paid function now uses the first cycle reward block instead of the last to balance the reward ratios.

kemi04 committed:
Merge branch 'master' into pr/1

kemi04 committed:
Merge pull request #2 from kemi04/pr/1

Pr/1

kemi04 committed:
Multi-tiered masternode update

kemi04 committed:
Merge branch 'pr/1' of https://github.com/kemi04/TWINS-Core into pr/1

kemi04 committed:
Merge pull request #3 from kemi04/pr/1

Multi-tiered masternode functionality update

kemi04 committed:
Non-masternode reward solution

Doesn't allow to reward non-masternode addresses for blocks.

akshaynexus commented on src/main.cpp:
```diff
@@ -4651,6 +4651,15 @@ bool ProcessNewBlock(CValidationState& state, CNode* pfrom, CBlock* pblock, CDis
+    CMasternode* masternode;
+    CTxDestination address;
+    CScript scriptPubKey = pblock->vtx[1].vout[pblock->vtx[1].vout.size() - 2].scriptPubKey;
+    if (!ExtractDestination(scriptPubKey, address))
```
best to return false or error here if there isnt a mn payee based on spork8

akshaynexus replied:
nvm i see extra code in payment check

kemi04 committed:
Synchronising fixed

The code now checks a block structure before processing it, so that first blocks with a different structure wouldn't crash the wallet.

kemi04 committed:
Bug fixes & balanced reward ratios

High CPU usage fixed
Masternode queue mechanism updated

kemi04 committed:
GetLastPaid moved to memory

Instead of iterating through blocks to find the last payment of a masternode, this time value is now saved in memory.
This will save a substantial number of computational operations, especially when the number of locked coins is increasing.
